### PR TITLE
Support object types as default arguments

### PR DESCRIPTION
**Pull request by varunagrawal**
This PR will allow us to define default arguments for object types via the default constructor.

E.g.
```cpp
void setPose(const gtsam::Pose3& pose = gtsam::Pose3());
```

will assign the default value to `pose`.

### CODE REVIEW

**Review by gchenfc — CHANGES_REQUESTED** [review]
See comment about saving `()` into the default string rather than having a bool as an `Argument` class member variable.

Also, slightly unrelated, but,
I can understand that you had issue with the original super-overpowered rule i had copy/pasted from SO for default args that basically just allowed anything up until the `,`, but I feel like by manually defining every possible type of default arg, we are investing a lot more engineering effort and bugs than if we just let the user use whatever syntax they want (as long as it's valid c++, which will be caught during pybind compile).  We can rely on the c++ compiler during pybind compilation, rather than us (1) basically re-writing the c++ parser all just to check if the input is valid, then (2) copy/paste the default arg from the `.i` file into the `.cpp` file.  Why not just skip step 1?

In this case of default arguments, I feel that the engineering effort is not warranted and if we just let allow any syntax up until the `,` (only exception is we have to ignore commas inside e.g. parentheses, brackets, etc), then _any_ C++ syntax is automatically supported without any additional effort on our part.

Furthermore, pyparsing's error messages are often completely and utterly unhelpful in debugging "syntax" errors (parse error: unexpected character when parsing), whereas if we just pass the default args straight to the `.cpp` file via the "anything up to the comma" method, then the user gets nice gcc syntax error messages which is actually useful in debugging.

So in conclusion, I would propose that we revert back to an over-powered parsing rule that basically allows any expression as a default argument up until 
1) the comma or 
2) close parenthesis
Taking care to ignore closed expressions of parentheses, brackets, or curly braces.  Then copy/paste that entire expression into the `cpp` file for
`py::arg("argname") = $expression`

File: gtwrap/pybind_wrapper.py [hunk 1]
```diff
@@ -48,7 +48,8 @@ def _py_args_names(self, args_list):
             py_args = []
             for arg in args_list.args_list:
                 if arg.default is not None:
-                    arg.default = ' = {arg.default}'.format(arg=arg)
+                    arg.default = ' = {arg.default}{parens}'.format(
+                        arg=arg, parens="()" if arg.default_has_parens else '')
```
[the remark this answers is not in the document]
Comment (varunagrawal) [reply gtwrap/pybind_wrapper.py:52]: Good point. I had too much on my plate and didn't consider this, but that's what reviews are for.

**Comment by varunagrawal** [comment]
That's funny because my review comment for the original rule was:

```
Seems to be overkill. We could just capture everything after the = upto the , and have the compiler verify if the type info is correct?
```

My issue wasn't with the rule but with testing out the capabilities. The wrapper was lying dormant due to a lack of testing. It's only because we have unit tests now that there is confidence when adding new features. I am not a fan of sacrificing testability and readability in favor of quick solutions, that's what lead the wrapper to be in a poor state in the first place.

Clearly we agree, I'll update the default arg rule.

**Comment by varunagrawal** [comment]
@gchenfc I made an updated rule that takes everything verbatim but is tuned to allow commas in the template parameters and argument list of the default.

**Review by gchenfc — CHANGES_REQUESTED** [review]
Ahh, great we're on the same page!  Sorry that I think I was sleep deprived the week of that old and wasn't paying careful attention.  Sorry about that :/

Suggested additional test cases:
```c++
std::vector<size_t> l = {1, 2},
gtsam::KeyFormatter lambda = [&c1, &c2](string s=5, int a){return s+"hello"+a+c1+c2;},
```

Suggested simplified implementation:
```python
DEFAULT_ARG = originalTextFor(
    OneOrMore(
        QuotedString('"') ^  #
        QuotedString("'") ^  #
        nestedExpr(opener='(', closer=')') ^  #
        nestedExpr(opener='[', closer=']') ^  #
        nestedExpr(opener='{', closer='}') ^  #
        nestedExpr(opener='<', closer='>') ^  #
        Word(printables, excludeChars="(){}[]<>,;")))
```
This way, no need to deal with Templates, Parameterized, numbers, etc. which introduce additional places for bugs to appear (btw, `originalTextFor` removes the need for `unquoteResults=False`).   IMHO, it's also more readable. :)

File: tests/test_interface_parser.py [hunk 2]
```diff
@@ -199,16 +195,31 @@ def test_default_arguments(self):
         self.assertEqual(args[6].default, 3)
         # No default argument should set `default` to None
         self.assertIsNone(args[7].default)
-
         self.assertEqual(args[8].default, 3.1415)
 
+        args = ArgumentList.rule.parseString("""
+            gtsam::KeyFormatter kf = gtsam::DefaultKeyFormatter,
+            std::vector<size_t> v = std::vector<size_t>(),
+            std::vector<size_t> l = {},
```
[the remark this answers is not in the document]
Comment (varunagrawal) [reply tests/test_interface_parser.py:203]: Ahhh good catch. Super glad we got you on board this project :smile:

**Comment by varunagrawal** [comment]
This is fantastic! Thanks.

> Suggested simplified implementation:
> 
> ```python
> DEFAULT_ARG = originalTextFor(
>     OneOrMore(
>         QuotedString('"') ^  #
>         QuotedString("'") ^  #
>         nestedExpr(opener='(', closer=')') ^  #
>         nestedExpr(opener='[', closer=']') ^  #
>         nestedExpr(opener='{', closer='}') ^  #
>         nestedExpr(opener='<', closer='>') ^  #
>         Word(printables, excludeChars="(){}[]<>,;")))
> ```
> 
> This way, no need to deal with Templates, Parameterized, numbers, etc. which introduce additional places for bugs to appear (btw, `originalTextFor` removes the need for `unquoteResults=False`). IMHO, it's also more readable. :)

File: tests/test_interface_parser.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (varunagrawal) [reply tests/test_interface_parser.py:203]: Also, I don't want to support lambdas as default arguments. They seem counterintuitive since one of the points of a lambda is to provide an easy one-off.

**Comment by varunagrawal** [comment]
> ```python
> DEFAULT_ARG = originalTextFor(
>     OneOrMore(
>         QuotedString('"') ^  #
>         QuotedString("'") ^  #
>         nestedExpr(opener='(', closer=')') ^  #
>         nestedExpr(opener='[', closer=']') ^  #
>         nestedExpr(opener='{', closer='}') ^  #
>         nestedExpr(opener='<', closer='>') ^  #
>         Word(printables, excludeChars="(){}[]<>,;")))
> ```

Does this work? I am getting all sorts of new errors in the tests.

File: tests/test_interface_parser.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gchenfc) [reply tests/test_interface_parser.py:203]: K, tbh I mainly just wanted a test that used a comma inside square brackets and a lambda was the first language feature I could think of that used that in a literal expression 😂

**Comment by gchenfc** [comment]
> > ```python
> > DEFAULT_ARG = originalTextFor(
> >     OneOrMore(
> >         QuotedString('"') ^  #
> >         QuotedString("'") ^  #
> >         nestedExpr(opener='(', closer=')') ^  #
> >         nestedExpr(opener='[', closer=']') ^  #
> >         nestedExpr(opener='{', closer='}') ^  #
> >         nestedExpr(opener='<', closer='>') ^  #
> >         Word(printables, excludeChars="(){}[]<>,;")))
> > ```
> 
> Does this work? I am getting all sorts of new errors in the tests.

It worked fine for me?  Oh sorry I forgot to mention that I had to quote all the ints and stuff in interface parser

e.g.
```python
        # self.assertEqual(args[1].default, 0)
        # self.assertEqual(args[2].default, 0)
        self.assertEqual(args[1].default, '0')
        self.assertEqual(args[2].default, '0.0')
```

were these the only errors or there were more?

I can push my changes if you want?


By the way, this will not explicitly ban lambdas, but doesn't require any extra effort to support lambdas.

**Comment by varunagrawal** [comment]
There are other issues with the rule: It allows things like `int x = <>` which shouldn't be allowed at the first pass rather than wait for long compile time cycles?

**Comment by varunagrawal** [comment]
While the new rule looks clean is it more readable though? I am not sure if someone else can easily parse what the rule is doing from a first glance. That was one of my objectives with breaking down the rule into constituents parts.

**Comment by gchenfc** [comment]
Ultimately, this is a matter of opinion.  I thought we had the same opinion with your quote 

> We could just capture everything after the = upto the , and have the compiler verify if the type info is correct"

but I guess not.  I strongly believe a more lenient wrapping rule is warranted, and I weakly believe the one I proposed is very reasonable.  Mainly, I just feel like manually dealing with all the random cases like templates, default args, constructors, initializer-lists, tuples - we're never going to get as good as the c++ compiler so why try to reinvent the wheel?  Just catch everything up until `,` and give it to the c++ compiler.

## Us catching things in the "first pass" (interface parser) rather than relying on the c++ compiler to catch them for us

> There are other issues with the rule: It allows things like `int x = <>` which shouldn't be allowed at the first pass rather than wait for long compile time cycles?

1) The current parsing rule also allows this, but that's a minor detail (see below for more examples of default args that are not caught properly by the current parser)
2) I would argue that readable error messages will save way more time than faster compiles, since you'll spend way longer debugging if the error message is totally useless than if the compile takes a couple minutes longer but gives you useful information.  Take the following example:
    ```
    File "/home/gerry/miniconda3/lib/python3.8/site-packages/pyparsing.py", line 3814, in parseImpl
        raise ParseException(instring, loc, self.errmsg, self)
    pyparsing.ParseException: Expected stringEnd, found 'n'  (at char 188), (line:8, col:1)
    ```
    when the real error was actually on line 512 of the .i file: `..., const int t = <,>);`.  Our parsing doesn't even correctly tell where in the file the error is!  Embarassing!
  with a much more lenient wrap parsing rule that allows almost anything up to the `,` or `;`, the c++ compiler throws this error:
    ```
    /home/gerry/GIT_REPOS/GTDynamics/build/python/gtdynamics.cpp: In function ‘void pybind11_init_gtdynamics(pybind11::module&)’:
    /home/gerry/GIT_REPOS/GTDynamics/build/python/gtdynamics.cpp:380:237: error: expected primary-expression before ‘<’ token
      380 |  result, t);}, py::arg("robot"), py::arg("result"), py::arg("t") = <,>)
          |                                                                    ^
    ```
    and it's super easy to tell exactly where and what the problem is.
3) Since the only thing we're discussing is syntax errors (pyparsing isn't going to catch things beyond syntax errors anyway), also keep in mind that syntax errors get caught quite quickly by a c++ compiler.  Admittedly, large header files with lots of template metaprogramming do still slow this down, but the point is that syntax errors get caught pretty early in a compile cycle.  Even in gtsam.cpp and adding a syntax error almost at the end of the file (`<,>`), it took <3min for my computer to throw the error very similar to the one quoted in bullet point 2.
4) people _much smarter than us_ have invested orders of magnitude _more time than us_ to make really nice c++ compilers that do all the syntax checking for us.  Why should we try to reinvent the wheel?

The current rules are definitely not any better (I would argue worse) at catching syntax errors than the proposed parsing rule.  Take the following examples:
```python
  args = ArgumentList.rule.parseString("""
      int a = {,
      int b = :,
      int c = o <, }{})([* , >(,*^&?),
      int d = 3.14,
      int e = globalVar.property
  """)[0].args_list
  for arg in args:
      print('{} = {}'.format(str(arg), arg.default))
```
output:
```
int a = {
int b = :
int c = o <, }{})([* , >(,*^&?)
int d = 3.14
int e = globalVar
```
note that `e` isn't parsed properly - it cuts off "property".

## Readability

> While the new rule looks clean is it more readable though? I am not sure if someone else can easily parse what the rule is doing from a first glance. That was one of my objectives with breaking down the rule into constituents parts.

Sure, readability is subjective.  IMO this rule has a reasonably clear english translation:
```python
DEFAULT_ARG = originalTextFor(
    OneOrMore(
        QuotedString('"') ^                   # allow quoted strings
        QuotedString("'") ^                   # allow quoted strings
        nestedExpr(opener='(', closer=')') ^  # allow matched (...) expressions
        nestedExpr(opener='[', closer=']') ^  # allow matched [...] expressions
        nestedExpr(opener='{', closer='}') ^  # allow matched {...} expressions
        nestedExpr(opener='<', closer='>') ^  # allow matched <...> expressions
        Word(printables, excludeChars="(){}[]<>,;")))  # stop at , or ; and also don't allow unmatched ()[]{}<>
```

_We don't need to know whether something is a template or basic type or anything like that_ - all we need to know is that we're stopping at `,` or `;`, except when they appear inside matched/nested expressions (we could actually use `nestedExpr(opener='"', closer='"')` instead of `QuotedString('"')` to make it more obvious what the rule is saying)

Maybe these comments might help:
```python
# Default argument passed to functions/methods.  Allow anything up to ',' or ';'
# except when they appear inside matched expressions such as
# (a, b) {c, b} "hello, how are you?"
DEFAULT_ARG = originalTextFor(
    OneOrMore(
        Word(printables, excludeChars="(){}[]<>,;") ^ # stop at , or ; unless they...
        QuotedString('"') ^                   # appear in "quoted strings"
        QuotedString("'") ^                   # appear in 'quoted strings'
        nestedExpr(opener='(', closer=')') ^  # appear in matched (...) expressions
        nestedExpr(opener='[', closer=']') ^  # appear in matched [...] expressions
        nestedExpr(opener='{', closer='}') ^  # appear in matched {...} expressions
        nestedExpr(opener='<', closer='>')    # appear in matched <...> expressions
    ))
```

Even though your code has broken up the rules into different pieces, I don't really find each individual piece very readable.  For example, it's not immediately clear what `{}:` supposed to be for in `BASIC_ARG = Word(alphas + "{}:")`  - why does a basic arg have curly braces and colons, possibly unmatched?  Sure, the final expression for `DEFAULT_ARG` is clear, but what about all the things that led up to it?

**Comment by gchenfc** [comment]
I've pushed what I propose and, at least on my machine, it passes all the unit tests.  Feel free to revert it if you don't like it.

**Comment by varunagrawal** [comment]
I was not doubting the fact that we should let the compiler handle type checks (I am very familiar with all the arguments you've made :) ), instead I was posing the question whether we should provide some sort of sanity checking at a basic level.

The readability question is legit though. Those curly braces in BASIC_ARG was for the vector initializer, but those should be removed since we added a separate rule for those. At this point, my only concern is that we are familiar with pyparsing so wrap is seeing activity, but what happens to the next person coming in? Is everything as clear and obvious?

**Comment by varunagrawal** [comment]
The poor error messages from pyparsing are something inherent to pyparsing, and there is a whole webpage dedicated by the author to explaining this. I have spent a lot of time trying to figure a way to print more useful error messages during parsing, so in the off chance we have a breakthrough, providing some meaningful errors seemed warranted. :sweat_smile: 

I personally felt the various components of `DEFAULT_ARG` were easy to parse individually and could be atomically tested which is why I favored that scheme. However your rule works well, we are on the same page of letting the compiler do the heavy lifting, and I see no reason to push my argument since it doesn't have any immediate obvious benefits. :slightly_smiling_face:

**Review by gchenfc — APPROVED** [review]
